### PR TITLE
[0.11] Restrict the number of maximal apps via configuration.

### PR DESCRIPTION
**Pull request by aquamatthias**
After merge: cherry pick on master!

### CODE REVIEW

**Comment by kolloch** [comment]
changelog is missing

File: docs/docs/command-line-flags.md [hunk 1]
```diff
@@ -125,7 +125,9 @@ The core functionality flags can be also set by environment variable `MARATHON_O
     automatically by Marathon.
     _Note: This prefix will not be added to variables that are already prefixed,
     such as `MESOS_TASK_ID` and `MARATHON_APP_ID`
-
+* <span class="label label-default">v0.11.0</span> `--max_apps` (Optional. Default: 500):
+    The number of applications, that are allowed to create.
```
Comment (kolloch) [inline docs/docs/command-line-flags.md]: The maximum number of applications that may be created.

File: src/main/scala/mesosphere/marathon/MarathonConf.scala [hunk 2]
```diff
@@ -196,4 +196,10 @@ trait MarathonConf
     validate = Set("zk", "mesos_zk", "mem").contains,
     default = Some("zk")
   )
+
+  lazy val maxApps = opt[Int]("max_apps",
+    descr = "The maximum number of apps allowed before Marathon will disallow new app creation.",
```
Comment (kolloch) [inline src/main/scala/mesosphere/marathon/MarathonConf.scala]: The same text as above?

File: src/main/scala/mesosphere/marathon/api/v2/ModelValidation.scala [hunk 3]
```diff
@@ -21,8 +21,17 @@ object ModelValidation {
...
     group: Group,
     path: String,
-    parent: PathId): Iterable[ConstraintViolation[V2Group]] =
-    checkGroup(V2Group(group), path, parent)
+    parent: PathId,
+    maximum: Int): Iterable[ConstraintViolation[V2Group]] = {
+
+    val v2Group = V2Group(group)
+    val maximumApps = if (group.transitiveApps.size > maximum) {
+      Seq(violation(v2Group, "apps", "", s"This Marathon instance may only handle up to $maximum Apps!"))
```
Comment (kolloch) [inline src/main/scala/mesosphere/marathon/api/v2/ModelValidation.scala]: It would be great to mention the config option here.

**Comment by kolloch** [comment]
Tests would be great.

**Comment by aquamatthias** [comment]
Thanks for the points. Hope I addressed them all.

File: src/main/scala/mesosphere/marathon/api/v2/ModelValidation.scala [hunk 4]
```diff
@@ -21,8 +21,18 @@ object ModelValidation {
   def checkGroup(
     group: Group,
     path: String,
-    parent: PathId): Iterable[ConstraintViolation[V2Group]] =
-    checkGroup(V2Group(group), path, parent)
+    parent: PathId,
+    maximum: Int): Iterable[ConstraintViolation[V2Group]] = {
```
Comment (kolloch) [inline src/main/scala/mesosphere/marathon/api/v2/ModelValidation.scala]: Sorry for not noting that before. I'd like some like `maxApps` better than `maximum`

File: src/test/scala/mesosphere/marathon/api/v2/ModelValidationTest.scala [hunk 5]
```diff
@@ -24,6 +24,21 @@ class ModelValidationTest
...
   }
 
+  test("A group can not be updated to have more than the configured number of apps") {
+    val group = Group("/".toPath, Set(
+      createServicePortApp("/a".toPath, 0).toAppDefinition,
+      createServicePortApp("/b".toPath, 0).toAppDefinition,
+      createServicePortApp("/c".toPath, 0).toAppDefinition
+    ))
+
+    val validations = ModelValidation.checkGroup(group, "", PathId.empty, 2)
```
Comment (kolloch) [inline src/test/scala/mesosphere/marathon/api/v2/ModelValidationTest.scala]: it would be nice to use a named argument for the number

**Comment by kolloch** [comment]
The changelog is still missing. Or do I miss something?

File: src/main/scala/mesosphere/marathon/api/v2/ModelValidation.scala (hunk 4 above)
Comment (aquamatthias) [inline src/main/scala/mesosphere/marathon/api/v2/ModelValidation.scala]: I agree.

**Comment by aquamatthias** [comment]
Added entry for changlog under 0.11 section.

**Comment by kolloch** [comment]
Cool! Waiting for tests. We can merge it then.

**Pull request merged**